datafusion/functions-array/src/position.rs

```diff
@@ -0,0 +1,400 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! [`ScalarUDFImpl`] definitions for array_position function.
+
+use arrow_schema::DataType::{FixedSizeList, LargeList, List, UInt64};
+use arrow_schema::{DataType, Field};
+use datafusion_common::plan_err;
+use datafusion_expr::expr::ScalarFunction;
+use datafusion_expr::Expr;
+use datafusion_expr::{ColumnarValue, ScalarUDFImpl, Signature, Volatility};
+use std::any::Any;
+use std::sync::Arc;
+
+use arrow_array::types::UInt64Type;
+use arrow_array::{
+    Array, ArrayRef, BooleanArray, GenericListArray, ListArray, OffsetSizeTrait, Scalar,
+    UInt32Array, UInt64Array,
+};
+use datafusion_common::cast::{
+    as_generic_list_array, as_int64_array, as_large_list_array, as_list_array,
+};
+use datafusion_common::{exec_err, internal_err};
+use itertools::Itertools;
+
+make_udf_function!(
+    ArrayPosition,
+    array_position,
+    array element index,
+    "searches for an element in the array, returns first occurrence.",
+    array_position_udf
+);
+
+#[derive(Debug)]
+pub(super) struct ArrayPosition {
+    signature: Signature,
+    aliases: Vec<String>,
+}
+impl ArrayPosition {
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::array_and_element_and_optional_index(
+                Volatility::Immutable,
+            ),
+            aliases: vec![
+                String::from("array_position"),
+                String::from("list_position"),
+                String::from("array_indexof"),
+                String::from("list_indexof"),
+            ],
+        }
+    }
+}
+
+impl ScalarUDFImpl for ArrayPosition {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+    fn name(&self) -> &str {
+        "array_position"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, arg_types: &[DataType]) -> datafusion_common::Result<DataType> {
+        Ok(match arg_types[0] {
```

Comment:
I don't think the type checking is necessary in `return_type` 🤔

Reply:
Addressed by reflecting both `array_position` and `array_positions`.